Move utility functions out of basic.py

samdoran opened the pull request:
##### SUMMARY
Move the following methods to `lib/anisble/module_utils/common/validation.py`:
- `_count_terms()`
- `_check_mutually_exclusive()`
- `_check_required_one_of()`
- `_check_required_together()`
- `_check_required_by()`
- `_check_required_arguments()`
- `_check_required_if`
- `fail_on_missing_params()` (create `check_missing_parameters()`)

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
`lib/ansible/module_utils/basic.py`
`lib/ansible/module_utils/common/validation.py`

samdoran committed:
Move _count_terms() out of basic

- Add unit tests

samdoran committed:
Update documentation

webknjaz commented on test/units/module_utils/common/test_validation.py:
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# (c) 2012-2014, Michael DeHaan <michael.dehaan@gmail.com>
+# (c) 2016 Toshio Kuratomi <tkuratomi@ansible.com>
+# (c) 2017-2018 Ansible Project
```
@samdoran these all lines are probably copy-pasted and need to be updated...

webknjaz commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -0,0 +1,22 @@
+    for term in check:
+        if term in params:
+            count += 1
+    return count
```
```suggestion
    return len(t for t in check if t in params)
```

webknjaz replied:
or
```suggestion
    return len(set(check.keys()) & set(params))
```

samdoran committed:
Update tests

samdoran committed:
Use set method rather than loop

Add unit tests

samdoran committed:
Move type checking into the function rather than calls

Update unit tests to reflect change in behavior

samdoran committed:
Move _check_mutually_exclusive() out of basic.py

Add unit tests

samdoran committed:
Move _check_required_one_of() out of basic.py

samdoran committed:
Change parameter variable from check to terms

samdoran committed:
Move _check_required_together() out of basic.py

Remove conditional in since an empty list has the same effect.

samdoran committed:
Move _check_required_by() out of basic.py

samdoran committed:
Move _check_required_arguments() out of basic.py

samdoran committed:
Move _check_required_if() out of basic.py

samdoran committed:
Move fail_on_missing_params() out of basic.py

This isn't really moving so much as creating a new function that returns a list of missing parameters. fail_on_missing_params() is not named as a private method, and moving it wholesale doesn't make much sense because fail_json() still lives in basic.py.

sivel commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -0,0 +1,242 @@
+
+
+def check_mutually_exclusive(terms, module_parameters):
+    """Check mutually exclusive terms against argument parameters. Accepts
```
I am split on my feelings about this function.  It doesn't work the same as it did as part of `AnsibleModule`.

Before, it would error on the first time that `count > 1`, whereas we aren't doing so now, we are collecting them all, but then not using them all.

Additionally, this is more like `list_mutually_exclusive`.  I think this function should raise some form of Exception, instead of just returning a list of the mutually exclusive args that were provided by the user.

samdoran replied:
> Before, it would error on the first time that count > 1, whereas we aren't doing so now, 

I wanted the new check method to return complete results which may be useful to the caller. I will change how this output is presented in `fail_json()` so that we use all the results and give a better error message. That is a change in behavior, but I think it's for the better.

> I think this function should raise some form of Exception, instead of just returning a list of the mutually exclusive args that were provided by the user.

Agreed. I will create an `AnsibleModuleParameterError` exception that contains the results of the failed check.

samdoran replied:
I implemented a custom error class (b371dc7c28), then just did it using `TypeError` and passing an already formatted message (f59cb9c9d0). I think starting simple with a built-in error class is the best approach. We can later subclass it if we want to add more parameters to the raised exception.

sivel commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -0,0 +1,242 @@
+
+
+def check_required_one_of(terms, module_parameters):
+    """Check each list of terms to ensure at least one exists in the given module
```
The same comment applies here.  I question the name vs what this is doing.  I feel like these `check_*` methods should be raising exceptions that are then caught in their `AnsibleModule` counterparts.  As it is, these are acting as "listing" functions, that then require the caller to know what a failure is based on what is returned.

As I am typing this I am still somewhat struggling, as it may be useful for the caller to also have the list.

I think in the end what I would like to see is an exception to indicate the failure, but with access in the exception to the list.

samdoran replied:
I changed all these functions to raise an exception and pass the results as a parameter (along with the original parameters as well, though we probably don't need those) to the exception. That way it can be used by the caller catching the exception.

sivel reviewed (changes requested):
I feel like overall this needs a small tweak.  My opinion is that the check functions are now really list functions that push the requirement of understanding what a failure is to the caller.  I believe we should be raising exceptions from the check functions, to clearly indicate a failure.

sivel commented:
I suppose if we were to rename these so that they were clearly listing functions, maybe that would solve my concerns.  I suppose my concern is around the fact that we have previously indicated that these `check_` type functions resulted in errors, where now they don't necessarily do so.

samdoran commented on lib/ansible/module_utils/common/messages.py:
```diff
@@ -0,0 +1,27 @@
+        })
+        self.log('[DEPRECATION WARNING] %s %s' % (msg, version))
+    else:
+        raise TypeError("deprecate requires a string not a %s" % type(msg))
```
Ignore this file. Wasn't supposed to include that.

Commit:
Change check_ functions to raise exceptions

- Create AnsibleModuleParameterError exception class for holding parameters that can be used by the caller catching the exception
- Create lib/ansible/module_utils/common/errors.py for holding custom error classes and move AnsibleModuleError here
- Change slightly the failure message of check_mutually_exclusive() to include all parameters not just the first ones. Adjust unit tests accordingly.

Commit:
Use TypeError rather than custom exception class
